internal/terraform/transform_test_run.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package terraform
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/hashicorp/terraform/internal/configs"
+	"github.com/hashicorp/terraform/internal/dag"
+	"github.com/hashicorp/terraform/internal/moduletest"
+)
+
+type TestRunTransformer struct {
+	File *moduletest.File
+}
+
+func (t *TestRunTransformer) Transform(g *Graph) error {
+	prevs := make(map[string]*NodeTestRun)
+	for _, run := range t.File.Runs {
+		node := &NodeTestRun{run: run, file: t.File}
+		g.Add(node)
+		refs, _ := run.GetReferences()
+		for _, ref := range refs {
+			subjectStr := ref.Subject.String()
+			if !strings.HasPrefix(subjectStr, "run.") {
+				continue
+			}
+			runName := strings.TrimPrefix(subjectStr, "run.")
+			if runName == "" {
+				continue
+			}
+			dependency, ok := prevs[runName]
+			if !ok {
+				// TODO: should we catch this error, or leave it, as it will still be caught in the test?
+				return fmt.Errorf("dependency %q not found for run %q", runName, run.Name)
```

Comment:
one of the advantages of moving to the graph is exactly this - we can catch these kind of errors right at the beginning and then report this as a kind of compilation error. So definitely leave this as is.